### DISCUSSION TITLE
CVU: New Deployment Tool For Yolov5

### DISCUSSION THREAD

**Discussion by shivamswarnkar** [ultralytics/yolov5#4291] [Show and tell]
Hi Everyone <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15">!

@aditya-dl  and I have been working on a pip package named [CVU-Python](https://github.com/BlueMirrors/cvu) <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15">  that can help you deploy Yolov5 models in various frameworks and devices, without much hassle.  We recently released its alpha-version and we would love to hear your feedback and suggestions on it.

Currently, we are giving support for GPU, CPU, and TPU devices through TensorRT, Torch-Script, ONNX, TensorFlow, and TFLite. And to switch between these devices and frameworks, all you need to do is to set ```device``` and ```backend``` parameters. CVU <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15"> even tries to take care of most of the installation requirements (even for TensorRT, tho we have only tested on Colab yet).

We are designing CVU <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15"> to be used by both the experts and the novices. CVU <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15"> aims at making CV pipelines easier to build and consistent around platforms, devices, and models.

We plan to extend CVU <img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="15"> to include other common computer vision tasks (such as tracking with deep sort) in the upcoming days!

We are also making submissions in the Yolov5 Export challenge, so please also check that out as well ([submission links](https://github.com/BlueMirrors/cvu#-cvu-computer-vision-utils)). 

Thank You all! Looking forward to hearing from you guys! 

## GPU (Colab-NVIDIA T4)

Based on 5000 inference iterations after 50 iterations of warmups. Includes Image Preprocessing (letterboxing etc.), Model Inference and Output Postprocessing (NMS, Scale-Coords, etc.) time.

| Backend    | FPS     |
| ---------- | ------- |
| TensorRT   | 157-165 |
| Torch      | 150-155 |
| ONNX       | 92-96   |
| TensorFlow | 43-47   |

## CPU (Colab)

Note: We are performing more benchmarks, we will update info later on.

Based on 500 inference iterations after 10 iterations of warmups. Includes Image Preprocessing (letterboxing etc.), Model Inference and Output Postprocessing (NMS, Scale-Coords, etc.) time.

| Backend    | FPS     |
| ---------- | ------- |
| ONNX       | 5.4-7.6 |
| TFLite     | 4.0-5.1 |
| TensorFlow | 3.5-4.2 |
| Torch      | 3.2-4.0 |

<img src="https://raw.githubusercontent.com/BlueMirrors/cvu/master/static/logo.png" width="25">
<img src="https://media.giphy.com/media/IFYptmLmY2ABjdcpfI/giphy.gif">

**Comment by glenn-jocher** [comment]
@shivamswarnkar nice work on the documentation and profiling especially!

**Reply by aditya-dl** [reply]
@glenn-jocher glad that you like it. We are planning to build more features inside the package that might speed up the inference times. xD

**Reply by shivamswarnkar** [reply]
@glenn-jocher  thank you! Please let us know if you have any suggestions or recommendations.

**Comment by ai-niki** [comment]
@shivamswarnkar nice work!
Wanted to test your solution and have several questions:
- Is there any example showing how to load a custom-trained yolov5 model?
- How would you suggest deploying a web app with maximum QPS? I'm familiar with Flask but it gave a bad performance in my experience. I'd like to reach at least 50 queries per second.

**Reply by shivamswarnkar** [reply]
@pokidyshev Thank you!

- Yeah, but it's not very well documented (we're still working on the alpha version, so we haven't had enough time to document yet). You can find an example code for each backend (for example [torch-doc](https://github.com/BlueMirrors/cvu#torch)) where we also explain how to use your custom weights. TLDR; just use ```weight="path_to_custom_weights"``` argument
  
- I think first I would recommend choosing the best inference backend for your hardware (onnx, Tensorrt, etc.). And to make sure that you pre-load and warm up the models before processing queries. And check out fast-API if you want to try some alternatives to Flask. 

- In my experience, if you want high QPS, you would need to utilize batching (assuming you'd be utilizing GPUs) + avoid unnecessary network latency (for example if you are working with a live stream or video, instead of sending every frame individually, I would recommend to compress multiple frames/images into one chunk, or use sockets/streaming-protocols instead of the client-server model). Tensorflow Serving does provide a good solution but unfortunately, in our implementations, the TensorFlow backend performed significantly worse than other backends([performance-comparison](https://github.com/BlueMirrors/cvu#yolov5-benchmark-results)). If you implement all of this on your own, then that could take some time. 

p.s. we are still working on the batch support, and haven't released it yet.
